src/s3ql/backends/b2/b2_backend.py

```diff
@@ -0,0 +1,807 @@
+        for key in metadata.keys():
+            if not isinstance(key, str):
+                raise ValueError('dict keys must be str, not %s' % type(key))
+
+            value = metadata[key]
+            if (not isinstance(value, (str, bytes, int, float, complex, bool))
+                and value is not None):
+                raise ValueError('value for key %s (%s) is not elementary' % (key, value))
+
+            if isinstance(value, (bytes, bytearray)):
+                value = base64.b64encode(value)
+
+            buffer = ('%s: %s' % (repr(key), repr(value)))
+            buffer = urllib.parse.quote(buffer, safe='!@#$^*()=+/?-_\'"><\\| `.,;:~')
+
+            if len(buffer) < chunksize:
+                metadata_dict['meta-%03d' % info_header_count] = buffer
+                info_header_count += 1
+            else:
+                i = 0
+                while i * chunksize < len(buffer):
+                    k = 'meta-%03d' % info_header_count
+                    v = buffer[i * chunksize : (i + 1) * chunksize]
+                    metadata_dict[k] = v
+                    i += 1
+                    info_header_count += 1
+
+        # Backblaze B2 only allows 10 metadata headers (2 are used by format and md5)
+        assert info_header_count <= 8
+
+        metadata_dict['meta-format'] = 'raw2'
+        md5 = base64.b64encode(checksum_basic_mapping(metadata)).decode('ascii')
+        metadata_dict['meta-md5'] = md5
+
+        return metadata_dict
+
+    def _add_b2_metadata_to_headers(self, headers, metadata, chunksize=2048):
+        '''Add metadata to HTTP headers
+
+        Theoretically there is no limit for the header size but we limit to
+        2014 bytes just to be sure the receiving server will handle it.
+        '''
+
+        metadata_dict = self._create_metadata_dict(metadata)
+
+        for key in metadata_dict.keys():
+            header_key = '%s%s' % (self.info_header_prefix, key)
+            encoded_value = self._b2_url_encode(metadata_dict[key])
+
+            headers[header_key] = encoded_value
+
+    def _extract_b2_metadata(self, response, obj_key):
+        '''Extract metadata from HTTP response object'''
+
+        headers = CaseInsensitiveDict()
+        for k, v in response.headers.items():
+            # we convert to lower case in order to do case-insensitive comparison
+            if k.lower().startswith(self.info_header_prefix.lower() + 'meta-'):
+                headers[k] = self._b2_url_decode(v)
+
+        format_ = headers.get('%smeta-format' % self.info_header_prefix, 'raw')
+        if format_ != 'raw2': # Current metadata format
+            raise CorruptedObjectError('invalid metadata format: %s' % format_)
+
+        parts = []
+        for i in count():
+            part = headers.get('%smeta-%03d' % (self.info_header_prefix, i), None)
+            if part is None:
+                break
+            parts.append(part)
+
+        buffer = urllib.parse.unquote(','.join(parts))
```

Comment:
```suggestion
        buffer = urllib.parse.unquote(''.join(parts))
```

When you currently safe more than 2048 bytes into a metadata key, this will not work. It will stick a `,` into a place where it does not belong. See https://github.com/s3ql/s3ql/blob/d00446705963317a6db5706228de110411918f49/src/s3ql/backends/s3c.py#L371 - there S3QL adds the `,` directly in the metadata. You need to do that, too.

This of course breaks all your currently persisted metadata. One could add a special case here for your legacy data (try to use `''.join(parts)` first, then `','.join(parts)` when that fails) but I wouldn't want to see those legacy codes in S3QL and I think @Nikratio wouldn't either. 
I think it's better that you re-write all metadata of your existing S3QL filesystems with a little upgrade-script.

Reply:
> I think it's better that you re-write all metadata of your existing S3QL filesystems with a little upgrade-script.

Okay, that was what I was afraid of.. :wink: